Fixed sidebar w. scrollable icons

rossbulat opened the pull request:
* fixed sidebar showing menu icons at all times.
* vertical scrolling introduced where menu height > screen height 
* collapsable toggle arrows rotate instead of snap
* item hover effect and general icon sizing optimisations

Provided this is merged, next steps would be to hide / show an expanded mobile side bar.

Fixed positioning:
![fixedsidebar1](https://user-images.githubusercontent.com/13929023/54019182-878c6a80-41c5-11e9-838b-bd7c4738e3af.gif)

Scrollable menu
![fixedsidebar2](https://user-images.githubusercontent.com/13929023/54019495-4b0d3e80-41c6-11e9-83e7-bb060c1f9ba1.gif)

jacogr commented on yarn.lock:
```diff
@@ -72,12 +83,12 @@
-  version "7.3.0"
-  resolved "https://registry.yarnpkg.com/@babel/helper-builder-react-jsx/-/helper-builder-react-jsx-7.3.0.tgz#a1ac95a5d2b3e88ae5e54846bf462eeb81b318a4"
-  integrity sha512-MjA9KgwCuPEkQd9ncSXvSyJ5y+j2sICHyrI0M3L+6fnS4wMSNDc1ARXsbTfbb2cXHn17VisSnU/sHFTCxVxSMw==
+"@babel/helper-builder-react-jsx@^7.0.0":
```
None of these look right - it is all downgrades from what we have?

rossbulat replied:
Hmm interesting. I thought I merged the current master yarn.lock. Will take another look!

rossbulat replied:
So i have just replaced it with the current master yarn.lock file.

Stefie commented on packages/apps/src/SideBar/index.tsx:
```diff
@@ -19,27 +18,15 @@ import Item from './Item';
+    const { isCollapsed } = this.props;
 
     return (
       <div className={`apps--SideBar ${isCollapsed ? 'collapsed' : 'expanded'}`}>
```
Related to the comment on Apps.tsx. It should be enough to add the `'collapsed' : 'expanded'` classes once on the outermost Application wrapper.

rossbulat commented on packages/apps/src/Apps.tsx:
```diff
@@ -18,23 +19,56 @@ import SideBar from './SideBar';
+          collapse={this.collapse}
+          isCollapsed={this.state.isCollapsed}
+        />
+        <Wrapper className={classes('apps-Wrapper', `${isCollapsed ? `collapsed` : `expanded`}`, `theme--${settings.uiTheme}`)}>
```
Understood, I think the padding was the initial concern with this, but will try to make it work. I like the idea of an additional wrapper if necessary.

Stefie commented:
Hi @rossbulat , thanks for you contribution, like @jacogr  I really like the ideas!

I left a few comments. As a summary: 
I would undo the change of the html structure since it's not really needed for what you want to achieve and I would favour to stick to CSS flexbox for positioning instead of `position: fixed;``

rossbulat commented on packages/apps/src/SideBar/SideBar.css:
```diff
@@ -5,66 +5,93 @@
-  position: relative;
-  text-align: left;
-  transition: width 0.3s;
+  position: fixed;
```
Thank you, I am pretty clued up with flex box. 

Would a flex box scroll element scroll with the entire window though? My intent is to keep the sidebar stuck so the icons don't disappear as you're scrolling.

rossbulat replied:
My other concern is that a fixed element would be needed for the sidebar to be compatible as a mobile menu, whereby the entire sidebar will be overlaying the main content. Are there any other ways to maintain the sidebar's fixed position purely with CSS?

I was considering a `flexbox: column` setup inside the `position: fixed` element as a means to align the menu items. So think there is scope to use flex box in conjunction with fixed elements.

Stefie replied:
> Would a flex box element scroll with the entire window though? My intent is to keep the sidebar stuck so the icons don't disappear as you're scrolling.

Here's a stripped down example of a layout with scrollable sidebar with flexbox: https://codepen.io/anon/pen/PLmQVg#anon-login

> Are there any other ways to maintain the sidebar's fixed position purely with CSS
A nice way is `position: sticky;`. Browser support is not where it should be, yet, but there are good polyfills. And the best news: The framework we're using  actually has a polyfilled Component for that: https://react.semantic-ui.com/modules/sticky/

rossbulat replied:
Oh fantastic! Good call - I'll go ahead and integrate `position: sticky` 😎

rossbulat commented on packages/apps/src/SideBar/SideBar.css:
```diff
@@ -5,66 +5,93 @@
 
   .ui.vertical.menu {
+    flex-grow: 1;
+    height: 110vh;
```
The reasoning for the 110vh is that there is a UI bug whereby the -webkit-scrollbar shows when an element is in a transition. It kind of flashes until the transition is complete.

The thinking here was to extend the width and make parent element overflow-y: hidden so the scrollbar artefect is never seen.

Stefie replied:
Might be related to the fixed positioning. I'd try to switch back to flexbox first and then see if the bug still persists.

rossbulat replied:
Indeed, will check if `position: sticky`will resolve.

rossbulat commented on packages/apps/src/Apps.tsx:
```diff
@@ -18,23 +19,56 @@ import SideBar from './SideBar';
+    };
+  }
+
+  private collapse = (): void => {
```
Noted! 👍

rossbulat commented:
> Hi @rossbulat , thanks for you contribution, like @jacogr I really like the ideas!
> 
> I left a few comments. As a summary:
> I would undo the change of the html structure since it's not really needed for what you want to achieve and I would favour to stick to CSS flexbox for positioning instead of `position: fixed;``

Great, I have implemented the proposals and committing the updated project now.

Stefie commented:
Also, the Icon somehow moved to the right in the expanded view which makes it basically impossible to toggle the view without moving the mouse. The icon should stay in one position when toggling between the two views:

![chrome-capture](https://user-images.githubusercontent.com/125398/54113395-7e2f1800-43e8-11e9-805c-33c096faa710.gif)

rossbulat commented:
> Also, the Icon somehow moved to the right in the expanded view which makes it basically impossible to toggle the view without moving the mouse. The icon should stay in one position when toggling between the two views. I can't pin it down by quickly looking at it:

So in essence you'd like to prioritise this over alignment with the text and transition direction.

rossbulat commented:
> Can we revert this - [0beb642](https://github.com/polkadot-js/apps/commit/0beb6425c9902ae5c13026c93f55433ff7630c11)
> 
> Would like to keep it with the Polkadot highlight colors as it was, so no theme changes here.

Well the problem here is that there are now 2 highlights happening, the background colour and the text colour, which together looks a bit over the top. Substrate theme did not originally have any hover effects happening, so for consistency I'd opt for keeping the background hover. 

In any case I'd advise to remove the background effect for the polkadot theme in the event we keep pink text highlighting (the active link remains pink).

jacogr commented:
Yes, the "Substrate theme" is not really actually a theme - it doesn't really have any overrides, it rather is a "bare-bones". (It should actually become a theme, but as of now it is just default colors without any hassle)

With the "Polkadot theme" we actually tried to align with the Polkadot colors along the highlights and kept it that way. (It is not 100%, the buttons still are not great all-over since there is not a lot of info in the base, but it is improving)

Rather do less instead of over-the-top that requires a move away.

Stefie commented:
> So in essence you'd like to prioritise this over alignment with the text and transition direction.

It would be aligned with the rest of the icons and the logo, like it was before.
And yes, I'd say that being able to click the toggle element twice without moving the mouse is more important than transitioning it.

rossbulat commented:
> > So in essence you'd like to prioritise this over alignment with the text and transition direction.
> 
> It would be aligned with the rest of the icons and the logo, like it was before.
> And yes, I'd say that being able to click the toggle element twice without moving the mouse is more important than transitioning it.

What would be cool is if the entire row was clickable, rather than just the circle itself. This way you'd get the best of both worlds.

rossbulat commented:
> ...
> Rather do less instead of over-the-top that requires a move away.

Understood, think I found a good balance. It looked a little odd hovering over the non-text area of the button and the pink text popping up, so kept a very subtle background hover.

jacogr commented on packages/apps/src/SideBar/index.tsx:
```diff
@@ -65,31 +52,24 @@ class SideBar extends React.PureComponent<Props, State> {
       <div className='apps--SideBar-collapse'>
         <Button
-          icon={`angle double ${isCollapsed ? 'right' : 'left'}`}
+          icon={`angle double right`}
```
Small nitpick - This works with a simple `icon='angle double right'` as well (no template string indicators and no unneeded `{}'s`

rossbulat replied:
Amended

rossbulat commented:
Noticed a padding issue. Will commit again shortly.

rossbulat commented:
Ok, should be good to go.
